src/monitor/node_active_protocol.c

```diff
@@ -1121,7 +1121,32 @@ RemoveNode(AutoFailoverNode *currentNode)
 
 		if (primaryNode)
 		{
+			ReplicationState goalState = primaryNode->goalState;
+
 			(void) ProceedGroupState(primaryNode);
+
+			/*
+			 * When the removal of a secondary node has no impact on the
+			 * primary node state, we still need to change the replication
+			 * settings to adjust to the possibly new
+			 * synchronous_standby_names, so we force APPLY_SETTINGS in that
+			 * case.
+			 */
+			if (primaryNode->goalState == goalState &&
+				goalState != REPLICATION_STATE_APPLY_SETTINGS)
+			{
+				LogAndNotifyMessage(
+					message, BUFSIZE,
+					"Setting goal state of " NODE_FORMAT
+					" to apply_settings after removing standby " NODE_FORMAT
+					" from formation %s.",
+					NODE_FORMAT_ARGS(primaryNode),
+					NODE_FORMAT_ARGS(currentNode),
+					formation->formationId);
+
+				SetNodeGoalState(primaryNode,
+								 REPLICATION_STATE_APPLY_SETTINGS, message);
```

Comment:
Not all valid primary states have a transition defined to `APPLY_SETTINGS` -- this results in things like
```
16:10:57 199 FATAL fsm.c:514 pg_autoctl does not know how to reach state "apply_settings" from "join_primary"
16:10:57 199 ERROR service_keeper.c:461 Failed to transition to state "apply_settings", retrying...
```
We have a unit test waiting in the wings that exposes things like this; I'll clean it up and file a draft.

Reply:
Oh good point. Yeah let's add the unit tests in a PR, and with a fix. The fix is that we don't need to go to apply_settings when we're join_primary, because the setting is going to be reset on the transition to primary anyway, as it should.